### ISSUE TITLE
Tenant could not be identified on domain tenancy

### ISSUE THREAD

**Issue by azamatx** [archtechx/tenancy#464]
#### Describe the bug
I can't access central domain. It says: "Tenant could not be identified on domain tenancy".

#### Steps to reproduce
Installed fresh copy of Laravel. Installed Tenancy and configured according to the documentation. Created a custom Tenant model:
```
class Tenant extends BaseTenant implements TenantWithDatabase
{
    use HasDatabase, HasDomains;
}
```
Changed some configuration in config/tanancy.php :

```
return [
    'tenant_model' => \App\Tenant::class,
    'id_generator' => null,
    'domain_model' => Domain::class,

    'central_domains' => [
        'tenancy.uz',
    ],
    ...
]
```

I'm planning to use multiple databases and subdomains. My default Laravel route is not changed. There's my tenancy route:
```
Route::middleware([
    'web',
    InitializeTenancyBySubdomain::class,
    PreventAccessFromCentralDomains::class,
])->group(function () {
    Route::get('/', function () {
        return 'This is your multi-tenant application. The id of the current tenant is ' . tenant('id');
    });
});
```
When I create tenants, their subdomains (foo.tenancy.uz, bar.tenancy.uz) work normal. But the central domain (tenancy.uz) is not accessible (not working).

More: https://flareapp.io/share/x5MkboPk

#### Expected behavior
Central domain should open my central app where users can sign up and create tenants.

#### Your setup
 - Laravel version: 7
 - stancl/tenancy version: 3
 - Storage driver: mysql

**Comment by stancl** [comment]
Try clearing config cache. If that doesn't help, visit our Discord for help. This is not a bug.

**Comment by johnyang90** [comment]
I am getting this error as well. But I am able to access central domain, but not tenant domain. After clear config cache still the same, is there anything wrong?

**Comment by bogordesaincom** [comment]
i have same problem

**Comment by mreduar** [comment]
For those who still have the problem, make sure your central domain is without http/https at first.
If you use `InitializeTenancyBySubdomain` middleware, make sure that the `domains` table does not contain the full domain.
If you are using the entire domain then change the middleware to `InitializeTenancyByDomain`

**Comment by tonaflcastelan** [comment]
Hi, I have a similar issue when I run tests in domain

```
1) Tests\Feature\CategoryApiTest::testGetCategory
Stancl\Tenancy\Exceptions\TenantCouldNotBeIdentifiedOnDomainException: Tenant could not be identified on domain tenant.localhost
```

Other issue is when I test a show method the model binding is empty


Any idea? I have read the **Frictionless Testing Setup**

**Comment by stancl** [comment]
What middleware are you using @tonaflcastelan?

**Comment by tonaflcastelan** [comment]
@stancl 

I have this setting in **tenant.php**

```
Route::prefix('api/v1')->middleware([
    'api',
    InitializeTenancyByDomain::class,
    PreventAccessFromCentralDomains::class,
])->group(function ($api) {
    $api->post('login', LoginController::class);
    $api->apiResource('categories', CategoryController::class);
});
```
Whe I run **php artisan test** command fails in the tests for domain, the central tests pass successfully (I follow the docs in **Frictionless Testing Setup**)

![image](https://user-images.githubusercontent.com/20670825/105506282-836d7700-5c8f-11eb-8d34-6defdacdafcf.png)

If I access to url **`(http://bar.localhost/api/v1/categories)`** I can see the data

![image](https://user-images.githubusercontent.com/20670825/105506423-a6982680-5c8f-11eb-9b3a-2c6c517a42e9.png)

Howhever if I change the setting in the **tenant.php** like this

```
Route::group([
    'prefix' => '/{tenant}/api/v1',
    'middleware' => ['api', InitializeTenancyByPath::class],
], function ($api) {
    $api->post('login', LoginController::class);
    $api->apiResource('categories', CategoryController::class);
});

```

All the tests pass and the access point is like this **`http://bar.localhost/bar/api/v1/categories`** (I still can see the data)

But I prefer the first way without prefix **tenant**

**Comment by stancl** [comment]
Did your tenant get created in the test? Can you `dd(Tenant::all(), Domain::all())`?

**Comment by tonaflcastelan** [comment]
If I put this `dd(Tenant::all(), Domain::all());` in the **CategoryTest** the Domain return empty

![image](https://user-images.githubusercontent.com/20670825/105565789-353b9080-5cee-11eb-9520-6efe6aea4a31.png)

If I put on the api route in **tenant.php** I get the data from **MySQL DB**, for my example **bar** and **foo**

![image](https://user-images.githubusercontent.com/20670825/105565894-b004ab80-5cee-11eb-997d-cffca097ff64.png)
![image](https://user-images.githubusercontent.com/20670825/105565905-bc890400-5cee-11eb-99f6-80309e749eb2.png)

My TenantTestCase looks like this

![image](https://user-images.githubusercontent.com/20670825/105565939-fe19af00-5cee-11eb-883a-aa6f5b583d3e.png)

My TestCase

![image](https://user-images.githubusercontent.com/20670825/105566337-4508a400-5cf1-11eb-84f8-3190652e0a25.png)


The **tenant.php** looks like this, with this code tests run succesfully, but If I uncomment below code the test must fail

![image](https://user-images.githubusercontent.com/20670825/105566055-af204980-5cef-11eb-9b54-8383834810f3.png)

And an example test

![image](https://user-images.githubusercontent.com/20670825/105566066-be9f9280-5cef-11eb-9923-84b30c195778.png)

And finally my **phpunit.xml config**, I use sqlite for the tests

![image](https://user-images.githubusercontent.com/20670825/105566078-d119cc00-5cef-11eb-8541-f19b8c445abd.png)

**Comment by tonaflcastelan** [comment]
If I set my phpunit tu run in MySQL

![image](https://user-images.githubusercontent.com/20670825/105566219-9d8b7180-5cf0-11eb-99fc-e560a8eb1c56.png)

Tenant and Domain are empty, this is the code:

```
Route::prefix('api/v1')->middleware([
    'api',
    InitializeTenancyByDomain::class,
    PreventAccessFromCentralDomains::class,
])->group(function ($api) {
    dd(Tenant::all(), Domain::all());
    $api->post('login', LoginController::class);
    $api->apiResource('categories', CategoryController::class);
});
```

The result

![image](https://user-images.githubusercontent.com/20670825/105566255-c6136b80-5cf0-11eb-885b-4669e50f189a.png)

And the exception

![image](https://user-images.githubusercontent.com/20670825/105566293-007d0880-5cf1-11eb-9438-a6cf27802d60.png)

**Comment by stancl** [comment]
Your `createTenant()` method doesn't create a domain, so there's no domain that the tenant can be identified on. Create a domain for the tenant and it should work 👍🏻

**Comment by tonaflcastelan** [comment]
Yeah! It works! Thanks @stancl :+1: 

Another question, when I run tests in **sqlite/memory** this pass succesfully, when I change **phpunit.xml** file to use a MySQL config the tests fails (only tenant tests)

I get this message

`SQLSTATE[HY000] [1049] Unknown database 'test_foo_app' (SQL: select max(`batch`) as aggregate from `migrations`)`

I tried to run `php artisan config:clear`, `php artisan config:cache` but is the same message

It seems DB tenant creation doen't run

As comment, I use RefreshDatabase also in Tests Tenants

![image](https://user-images.githubusercontent.com/20670825/105616959-77bda580-5da0-11eb-941e-6866af47b171.png)

**Comment by stancl** [comment]
Make sure that you're using the right DB config in `config/tenancy.php`, e.g. the DB managers, connections, etc.

**Comment by tonaflcastelan** [comment]
The config in tenancy.php is default, for now I try with sqlite, in the future I will try run tests in MySQL

Thank you very much for your help

Awesome package

**Comment by tintucraju** [comment]
How can I add a custom error page in this situation. If tenant is not found show a 404 error page.

**Comment by stancl** [comment]
![BCB4D8F0-9ACE-4C62-89CE-FB8FC51607EF](https://user-images.githubusercontent.com/33033094/118607553-01aed580-b7b9-11eb-95a1-044962b0dc4f.jpeg)

**Comment by sarmenb** [comment]
What i did is in the Providers/TenancyServiceProvider in the boot method i put this and it works like a charm.

InitializeTenancyByDomain::$onFail = function ($exception, $request, $next) {
            return redirect('http://my-domain.com');
};

**Comment by osmanmoharram** [comment]
In the [quick start page](https://tenancyforlaravel.com/docs/v3/quickstart) go to centeral routes section and make modifications accordingly.

**Comment by sicaboy** [comment]
TLDR;
Solution: Just make sure your 
```
$this->tenant->getTenantKey()
```
returns something.

**Comment by SkyNetRu** [comment]
[> Your `createTenant()` method doesn't create a domain, so there's no domain that the tenant can be identified on. Create a domain for the tenant and it should work 👍🏻](https://github.com/archtechx/tenancy/issues/464#issuecomment-766150877)

This helps me, thanks a lot
